exatomic/va/va.py

```diff
@@ -0,0 +1,327 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2015-2018, Exa Analytics Development Team
+# Distributed under the terms of the Apache License 2.0
+"""
+Vibrational Averaging
+#########################
+Collection of classes for VA program
+"""
+import numpy as np
+import pandas as pd
+import csv
+import os
+from exa.util.units import Length
+from exa.util.utility import mkp
+
+_gauss_template='''{link0}
+{route}
+
+{title}
+
+{charge} {mult}
+'''
+float_format = '%    .8f'
+class GenInput:
+    """
+    Supporting class for Vibrational Averaging that will generate input files
+    for a selected program under a certain displacement parameter.
+
+    Computes displaced coordinates for all available normal modes from the equilibrium
+    position by using the displacement vector components contained in the
+    :class:`~exatomic.atom.Frequency` dataframe. It will scale these displacements to a
+    desired type defined by the user with the delta_type keyword. For more information
+    on this keyword see the documentation on the
+    :class:`~exatomic.va.va.GenInputs.gen_delta` function.
+
+    Args:
+        uni (:class:`~exatomic.Universe`): Universe object containg pertinent data
+        delta_type (int): Integer value to define the type of delta parameter to use
+    """
+
+    def _gen_delta(self, freq, delta_type):
+        """
+        Function to compute the delta parameter to be used for the maximum distortion
+        of the molecule along the normal mode.
+
+        When delta_type = 0 we normalize the displacments to have a maximum of 0.04 Bohr
+        on each normal mode.
+
+        When delta_type = 1 we normalize all atomic displacements along all normal modes
+        to have a global average displacement of 0.04 Bohr.
+
+        When delta_type = 2 we normalize each displacement so every atom has a maximum
+        displacement of 0.04 Bohr on every normal mode.
+
+        Args:
+            freq (:class:`exatomic.atom.Frequency`): Frequency dataframe
+            delta_type (int): Integer value to define the type of delta parameter to use
+        """
+        # average displacement of 0.04 bohr for each normal mode
+        nat = len(freq['label'].drop_duplicates())
+        nmode = len(freq['freqdx'].drop_duplicates())
+        freqdx = freq['freqdx'].values
+        if delta_type == 0:
+            # Code using for loops
+            # a = np.linalg.norm(freq[['dx', 'dy', 'dz']].
+            #                                    values, axis=1, ord=2)
+            # vec_sum = []
+            # for i in range(0,len(a),3):
+            #     vec_sum.append([0])
+            #     for j in range(i,i+3):
+            #         vec_sum[-1] += a[j]
+            # print(a)
+            # print(vec_sum)
+            d = freq.groupby(['freqdx', 'frame']).apply(
```

Comment:
An alternative approach to this could be to do something like
```
@numba.jit
def func(args):
    return sum(linalg)

freq.groupby([...])[['dx', 'dy', 'dz']].apply(func)
```
which as highlight opens the possibility for performance gains on the computation itself (not sure if that is the bottleneck here).

Reply:
The bottleneck is really the quantum chemistry calculations. This is being applied to a system with 121 atoms and there's not a big issue with this part of the code. It would benefit greatly if we could do less I/O iterations. It makes about 4*nmodes files which I have already worked on to only select a small list of modes.

Reply:
Okay understood. Keep in mind that we want to try to not repeat ourselves (keep it DRY as they say). Even if performance isn't the main impetus, code cleanliness, testability, and atomicity could benefit.

Reply:
I agree, I'll keep that in mind in the future.